### PR TITLE
Port fbstring_core to big-endian architectures.

### PR DESCRIPTION
**Pull request by cole14**
There's 2 ways this could be implemented - either as a series of preprocessor blocks depending on target architecture (as I have implemented it here), or by encapsulating access to MediumLarge::capacity_ within getters/setters as in a similar manner to setSmallSize() and smallSize(). The first option makes the code a bit harder to read, but the second option changes the existing control flow a bit which could slightly alter performance.

I opted for the first so as to keep the existing amd64 flow untouched, but can easily change the pull request to the second option to keep code readability a priority.

### CODE REVIEW

**Comment by chrismwendt** [comment]
The documentation at folly/docs/FBString.md might need to be updated as well.

**Comment by cole14** [comment]
Updated folly/docs/FBString.md to reflect that both big and little endian is now supported as per @chrismwendt

**Comment by Orvid** [comment]
It does appear that a good portion of the ifdefs could be abstracted out and centralized. That might make the code quite a bit cleaner.

**Comment by cole14** [comment]
Updated to combine most of the ifdefs into a single MediumLarge::setCapacity function, greatly improving code readability.

File: folly/docs/FBString.md [hunk 1]
```diff
@@ -9,8 +9,8 @@ allocator. In particular, `fbstring` is designed to detect use of
 jemalloc and cooperate with it to achieve significant improvements in
 speed and memory usage.
 
-`fbstring` supports x32 and x64 architectures. Porting it to big endian
-architectures would require some changes.
+`fbstring` supports 32- and 64-bit and little- and big-endian
```
Comment (Orvid) [inline folly/docs/FBString.md]: I don't typically see `32-bit` as `32-`, even when referenced immediately before `64-bit`. Normally it's just `32`. The same case can be made for `little-`.

**Comment by cole14** [comment]
Any update on this PR?

**Comment by JoelMarcey** [comment]
We had an internal review of this and @yfeldblum had a comment...

> Use `constexpr` throughout instead of macros.

> OK to define the `constexpr` once at the top in terms of the macro though if necessary. Clang doesn't allow `reinterpret_cast` in `constexpr` expressions, so tricks like testing byte 0 of a 2-byte word or this may not be possible.

**Comment by Orvid** [comment]
It's worth noting that, except in the root evaluation scope, which they don't document, MSVC doesn't allow `reinterpret_cast` in `constexpr` expressions either.

**Pull request closed without merging**